File: Euler_98.py
import sys
import math
import re
from sets import Set
from datetime import datetime

def find_unique_squares():
    square_sets = []
    square_is_anagram = False

    for i in range(100000):
        square = i**2

        if len(str(square)) == len(Set([c for c in str(square)])):
            for square_set in square_sets:
                if sorted(str(square)) == sorted(str(square_set[0])):
                    square_set.append(square)
                    square_is_anagram = True
                    break

            if not square_is_anagram:
                square_set = []
                square_set.append(square)
                square_sets.append(square_set)

            square_is_anagram = False

    squares = []

    for square_set in square_sets:
        if len(square_set) > 1:
            squares.append(square_set)

    return squares

def find_all_anagrams(words):
    anagram_sets = []
    word_is_anagram = False

    for word in words:
        for anagram_set in anagram_sets:
            if sorted(word) == sorted(anagram_set[0]):
                anagram_set.append(word)
                word_is_anagram = True
                break

        if not word_is_anagram:
            anagram_set = []
            anagram_set.append(word)
            anagram_sets.append(anagram_set)

        word_is_anagram = False

    anagrams = []

    for anagram_set in anagram_sets:
        if len(anagram_set) > 1:
            anagrams.append(anagram_set)

    return anagrams

with open('files/p098_words.txt') as f:
    words = re.split('","',f.read()[1:-1])

a = datetime.now()

anagram_sets = find_all_anagrams(words)
square_sets  = find_unique_squares()

delta = datetime.now() - a

print "total time: " + str(delta.total_seconds())

File: Euler_98.cpp
#include "Euler.h" 
#include <algorithm>
#include <fstream>
#include <vector>
#include <unordered_map>

int anagrams_match(std::vector<std::string>& anagram_pair, std::vector<std::string>& anagram_squares) {
    int solution = -1;

    for (uint64_t i = 0; i < anagram_squares.size(); ++i) {
        std::unordered_map<char, char> char_map;
        bool match = true;

        for (uint64_t c = 0; c < anagram_pair[0].size(); ++c) {
            bool char_in_map = char_map.find(anagram_pair[0][c]) != char_map.end();

            for (std::pair<char, char> p : char_map) {
                if (p.second == anagram_squares[i][c] && ((p.first != anagram_pair[0][c]) || !char_in_map)) {
                    match = false;
                    break;
                }
            }

            if (match && !char_in_map) {
                char_map.insert(std::pair<char, char>(anagram_pair[0][c], anagram_squares[i][c]));
            }
        }

        if (match) {
            std::string built;

            for (uint64_t c = 0; c < anagram_pair[1].size(); ++c) {
                built.append(&char_map.find(anagram_pair[1][c])->second);
            }

            for (std::string anagram_square : anagram_squares) {
                if (built == anagram_square) {
                    solution = std::max(std::max(std::stoi(built), std::stoi(anagram_squares[i])), solution);
                }
            }
        }
    }

    return solution;
}

std::vector<std::vector<std::string> > find_all_anagrams(std::vector<std::string>& words) {
    std::vector<std::vector<std::string> > anagram_pairs;

    for (std::string& word : words) {
        bool word_is_anagram = false;

        for (std::vector<std::string>& anagram_pair : anagram_pairs) {
            std::string temp1 = word; std::sort(temp1.begin(), temp1.end());
            std::string temp2 = anagram_pair[0]; std::sort(temp2.begin(), temp2.end());

            if (temp1 == temp2) {
                word_is_anagram = true;
                anagram_pair.push_back(word);
                break;
            }
        }

        if (!word_is_anagram) {
            anagram_pairs.push_back(std::vector<std::string>({word}));
        }
    }

    return anagram_pairs;
}

int Euler::AnagramicSquares() {
    std::vector<std::string> words, squares;
    std::ifstream file;
    std::string temp;

    file.open("files/p098_words.txt");

    while(std::getline(file, temp, ',')) {
        temp.erase(temp.size() - 1, 1); temp.erase(0, 1);
        words.push_back(temp);
    }

    file.close();

    for (int i = 0; i < sqrt(1e9); ++i) {
        squares.push_back(std::to_string(i * i));
    }

    std::vector<std::vector<std::string> > anagram_pairs = find_all_anagrams(words);
    std::vector<std::vector<std::string> > anagram_squares = find_all_anagrams(squares);

    int largest_square = 0;

    for (std::vector<std::string>& anagram_pair : anagram_pairs) {
        for (std::vector<std::string>& anagram_square : anagram_squares) {
            if (anagram_square.size() > 1 && anagram_square[0].size() == anagram_pair[0].size() && anagram_pair.size() > 1) {
                int solution = anagrams_match(anagram_pair, anagram_square);
                if (solution > largest_square) {
                    largest_square = solution;
                }
            }
        }
    }

    return largest_square;
}
